drivers/input/input_adc_keys.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (c) 2024 Chen Xingyu <hi@xingrz.me>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adc_keys
+
+#include <stdlib.h>
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+
+LOG_MODULE_REGISTER(adc_keys, CONFIG_INPUT_LOG_LEVEL);
+
+struct adc_keys_code_config {
+	uint32_t key_index;
+	uint32_t press_mv;
+	uint16_t zephyr_code;
+};
+
+struct adc_keys_config {
+	struct adc_dt_spec channel;
+	uint32_t sample_period_ms;
+	uint32_t keyup_mv;
+	const struct adc_keys_code_config *code_cfg;
+	uint32_t code_cnt;
+	uint32_t key_cnt;
+};
+
+struct adc_keys_key_data {
+	const uint16_t zephyr_code;
+	int32_t last_value;
+	int32_t curr_value;
+};
+
+struct adc_keys_data {
+	const struct device *self;
+	struct k_work_delayable dwork;
+	struct adc_sequence seq;
+	struct adc_keys_key_data *key_data;
```

Comment:
This pointer can go in the config structure since the pointer itself does not change, save few bytes of ram. Then at that point the only thing to initialize is `self`, you can set that in `adc_keys_init` and drop the static initialization, which saves few bytes of flash as well.

Reply:
Yeah that's right, but if you initialize it statically then it's going to take the space in RAM and also space in flash for the static initialization, including for the members that you are not explicitly initializing (storing a bunch of 0 essentially), while if you let it uninitialized the system will zero-initialize it automatically and you just use the flash for the instructions to set the pointer, so it's usually a good idea to leave data structures uninitalized if possible.